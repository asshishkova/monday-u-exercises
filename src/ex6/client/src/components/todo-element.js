import React, { useEffect, useState } from "react";
import { deleteItem, markItemAsOld, changeItemStatus } from "../item_client.js";

export function TodoElement(props) {
  const [todoClassName, setTodoClassName] = useState("todo-li existing-todo")

  const todo = props.todo;
  const doneTime = todo.done === null? "" : `Done at ${todo.done.slice(11,16)} ${todo.done.slice(0,10)}`;
  const checked = todo.status? "checked" : "";

  useEffect(() => {
    if (todo.isNew) {
      setTodoClassName("todo-li animation-add-todo");
      markItemAsOld(todo);
    }
  },[props, todo])

  const animationEndHandler = () => {
    if (todoClassName === "todo-li animation-add-todo") {
      setTodoClassName("todo-li existing-todo")
    } else {
      props.updateTodos();
    }
  }

  const onDeleteButtonClicked = async (e) => {
    await deleteItem(todo);
    setTodoClassName("todo-li animation-delete-todo");
    // updateTodos runs in animationEndHandler
  }

  const onCheckboxClicked = async (e) => {
    await changeItemStatus(todo);
    await props.updateTodos();
  }

  const todoItem =  <label className="todo-item" info={doneTime}>
                      <label className="todo-item-checkbox">
                        <div className="todo-item-text">{todo.text}</div>
                        <input type="checkbox" className="status-checkbox" defaultChecked={checked} onClick={onCheckboxClicked} />
                        <span className="status-checkbox-mark"></span>
                      </label>
                    </label>

  const deleteButton = <button className="delete-todo-button btn" onClick={onDeleteButtonClicked}>
                          <i className="fa fa-trash"></i></button>

  return (
    <li className={todoClassName} onAnimationEnd={animationEndHandler}>
      {todoItem}
      {deleteButton}
    </li>
  )
}
