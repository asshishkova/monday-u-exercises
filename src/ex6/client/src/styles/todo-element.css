@import "checkbox.css";

.todo-li {
  list-style-type: none;
  margin-top: 10px;
  border-radius: 4px;
  display: flex;
}

.todo-item {
  padding-right: 10px;
  width: 415px;
  min-height: 40px;
  cursor: pointer;
  display: flex;
  align-items: center;
  position: relative;
}

.todo-item-text {
  width: 370px;
  overflow: scroll;
}

.existing-todo {
  background-color: rgba(128, 128, 128, 0.2);
}

.existing-todo:hover {
  background-color: rgba(128, 128, 128, 0.4);
}

.animation-add-todo {
  animation: change-color 0.7s linear;
}

@keyframes change-color {
  0% {
    background: rgba(128, 128, 128, 0.8);
  }
  100% {
    background: rgba(128, 128, 128, 0.2);
  }
}

.delete-todo-button {
  font-family: "";
  background-color: brown;
  height: 40px;
  width: 45px;
  display: none;
}

.existing-todo:hover .delete-todo-button {
  display: block;
}

.animation-delete-todo {
  color: antiquewhite;
  animation: deleting-color 0.7s ease-out;
}

@keyframes deleting-color {
  0% {
    opacity: 1;
    background: brown;
  }
  100% {
    opacity: 0;
    background: antiquewhite;
  }
}

.todo-item[info]:hover::after, .todo-item[info]:after {
  content: attr(info);
  position: absolute;
  color: antiquewhite;
  background-color: gray;
  left: 35px;
  top: 30px;
  font-size: 14px;
  border-radius: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

.todo-item[info]:after {
  opacity: 0;
  transition: 0s all;
}
.todo-item[info]:hover::after {
  opacity: 1;
  transition-delay: 1.5s;
}
