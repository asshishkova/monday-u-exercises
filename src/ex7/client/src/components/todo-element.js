import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { getItems } from "../item-client.js";
import { deleteItem, markItemAsOld, changeItemStatus } from "../item-client.js";
import "../styles/todo-element.css";

const NO_ANIMATION = Symbol("no animation");
const ADDING_ANIMATION = Symbol("adding animation");
const DELETING_ANIMATION = Symbol("deleting animation");

export function TodoElement({ todo,
                              deleteTodoAction,
                              setServerErrorMessageAction,
                              saveDeletedItemAction,
                              setTodosAction,
                              markOldAction
                            }) {

  const [currentAnimation, setCurrentAnimation] = useState(NO_ANIMATION)

  const doneTime = todo.done === null? "" : `Done at ${todo.done.slice(11,16)} ${todo.done.slice(0,10)}`;
  const checkedDefaultValue = todo.status? "checked" : "";

  useEffect(() => {
    const markItemAsOldEffect = async () => {
      try {
        await markItemAsOld(todo);
        markOldAction(todo);
      } catch (error) {
        setServerErrorMessageAction(`Error: ${error.message}`);
      }
    }
    if (todo.isNew) {
      setCurrentAnimation(ADDING_ANIMATION);
      markItemAsOldEffect();
    }
  },[todo, setServerErrorMessageAction, markOldAction])

  const animationEndHandler = async (todo) => {
    setCurrentAnimation(NO_ANIMATION);
    if (currentAnimation === DELETING_ANIMATION) {
      deleteTodoAction(todo);
      saveDeletedItemAction(todo);
    }
  }

  const onDeleteButtonClicked = async (e) => {
    setServerErrorMessageAction("");
    try {
      await deleteItem(todo);
      setCurrentAnimation(DELETING_ANIMATION);
      // setTodosAction and saveDeletedItemAction runs in animationEndHandler
    } catch (error) {
      setServerErrorMessageAction(`Error: ${error.message}`);
    }
  }

  const onCheckboxClicked = async (e) => {
    setServerErrorMessageAction("");
    try {
      await changeItemStatus(todo);
      setTodosAction(await getItems());
    } catch (error) {
      setServerErrorMessageAction(`Error: ${error.message}`);
    }
  }

  let todoClassName;
  if (currentAnimation === NO_ANIMATION) {
    todoClassName = "todo-li existing-todo";
  } else if (currentAnimation === ADDING_ANIMATION) {
    todoClassName = "todo-li existing-todo animation-add-todo";
  } else { // if currentAnimation === DELETONG_ANIMATION
    todoClassName = "todo-li animation-delete-todo";
  }

  let todoItem;

  todoItem =  <label className="todo-item" info={doneTime}>
                <label className="todo-item-checkbox">
                  <div className="todo-item-text">{todo.text}</div>
                  <input type="checkbox" className="status-checkbox" defaultChecked={checkedDefaultValue} onClick={onCheckboxClicked} />
                  <span className="status-checkbox-mark"></span>
                </label>
              </label>

  const deleteButton = <button className="delete-todo-button btn" onClick={onDeleteButtonClicked}>
                          <i className="fa fa-trash"></i></button>

  return (
    <li className={todoClassName} onAnimationEnd={() => animationEndHandler(todo)} >
      { todoItem }
      { deleteButton }
    </li>
  )
}

TodoElement.propTypes = {
  todo: PropTypes.object,
}
