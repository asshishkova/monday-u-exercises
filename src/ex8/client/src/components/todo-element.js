import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { getItems } from "../item-client.js";
import { deleteItem, markItemAsOld, changeItemStatus } from "../item-client.js";
import "../styles/todo-element.css";

const NO_ANIMATION = Symbol("no animation");
const ADDING_ANIMATION = Symbol("adding animation");
const DELETING_ANIMATION = Symbol("deleting animation");

export function TodoElement({ todo,
                              deleteTodoAction,
                              addTodosAction,
                              setServerErrorMessageAction,
                              clearServerErrorMessageAction,
                              saveDeletedItemAction,
                              setTodosAction,
                              markOldAction }) {

  const [currentAnimation, setCurrentAnimation] = useState(NO_ANIMATION)

  const doneTime = todo.done === null? "" : `Done at ${todo.done.slice(11,16)} ${todo.done.slice(0,10)}`;
  const checkedDefaultValue = todo.status? "checked" : "";

  useEffect(() => {
    const markItemAsOldEffect = async () => {
      clearServerErrorMessageAction();
      try {
        await markItemAsOld(todo);
        markOldAction(todo);
      } catch (error) {
        setServerErrorMessageAction(`Error: ${error.message}`);
      }
    }
    if (todo.isNew) {
      setCurrentAnimation(ADDING_ANIMATION);
      markItemAsOldEffect();
    }
  },[todo, setServerErrorMessageAction, markOldAction, clearServerErrorMessageAction])

  const animationEndHandler = async (todo) => {
    setCurrentAnimation(NO_ANIMATION);
    if (currentAnimation === DELETING_ANIMATION) {
      deleteTodoAction(todo); // redux optimistic deleting
      saveDeletedItemAction(todo);
      try {
        await deleteItem(todo);
      } catch (error) {
        addTodosAction([todo]);
        saveDeletedItemAction(null);
        setServerErrorMessageAction(`Error: ${error.message}`);
      }
    }
  }

  const onDeleteButtonClicked = async (e) => {
    clearServerErrorMessageAction();
    setCurrentAnimation(DELETING_ANIMATION); // redux optimistic deleting
  }

  const onCheckboxClicked = async (e) => {
    clearServerErrorMessageAction();
    try {
      await changeItemStatus(todo);
      setTodosAction(await getItems());
    } catch (error) {
      setServerErrorMessageAction(`Error: ${error.message}`);
    }
  }

  let todoClassName;
  if (currentAnimation === NO_ANIMATION) {
    todoClassName = "todo-li existing-todo";
  } else if (currentAnimation === ADDING_ANIMATION) {
    todoClassName = "todo-li existing-todo animation-add-todo";
  } else { // if currentAnimation === DELETONG_ANIMATION
    todoClassName = "todo-li animation-delete-todo";
  }

  let todoItem;

  todoItem =  <label className="todo-item" info={doneTime}>
                <label className="todo-item-checkbox">
                  <div className="todo-item-text">{todo.text}</div>
                  <input type="checkbox" className="status-checkbox" defaultChecked={checkedDefaultValue} onClick={onCheckboxClicked} />
                  <span className="status-checkbox-mark"></span>
                </label>
              </label>

  const deleteButton = <button className="delete-todo-button btn" onClick={onDeleteButtonClicked}>
                          <i className="fa fa-trash"></i></button>

  return (
    <li className={todoClassName} onAnimationEnd={() => animationEndHandler(todo)} >
      { todoItem }
      { deleteButton }
    </li>
  )
}

TodoElement.propTypes = {
  todo: PropTypes.object,
}
