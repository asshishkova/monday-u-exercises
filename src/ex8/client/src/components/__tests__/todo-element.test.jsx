import renderer from "react-test-renderer";
import { TodoElement } from "../todo-element.js";

const pendingTodo = {id: 1, text: 'do this', status: false, done: null}
const doneTodo = {id: 2, text: 'already done this', status: true, done: "2022-07-13T12:49:57.000Z"}

test('renders correctly pending todo', () => {
  const tree = renderer.create(
    <TodoElement key={pendingTodo.id} todo={pendingTodo}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly done todo', () => {
  const tree = renderer.create(
    <TodoElement key={doneTodo.id} todo={doneTodo}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
