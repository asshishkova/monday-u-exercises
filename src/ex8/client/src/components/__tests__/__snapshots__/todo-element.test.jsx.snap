// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly done todo 1`] = `
<li
  className="todo-li existing-todo"
  onAnimationEnd={[Function]}
>
  <label
    className="todo-item"
    info="Done at 12:49 2022-07-13"
  >
    <label
      className="todo-item-checkbox"
    >
      <div
        className="todo-item-text"
      >
        already done this
      </div>
      <input
        className="status-checkbox"
        defaultChecked="checked"
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="status-checkbox-mark"
      />
    </label>
  </label>
  <button
    className="delete-todo-button btn"
    onClick={[Function]}
  >
    <i
      className="fa fa-trash"
    />
  </button>
</li>
`;

exports[`renders correctly pending todo 1`] = `
<li
  className="todo-li existing-todo"
  onAnimationEnd={[Function]}
>
  <label
    className="todo-item"
    info=""
  >
    <label
      className="todo-item-checkbox"
    >
      <div
        className="todo-item-text"
      >
        do this
      </div>
      <input
        className="status-checkbox"
        defaultChecked=""
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="status-checkbox-mark"
      />
    </label>
  </label>
  <button
    className="delete-todo-button btn"
    onClick={[Function]}
  >
    <i
      className="fa fa-trash"
    />
  </button>
</li>
`;
